queries/purescript/highlights.scm

```diff
@@ -0,0 +1,149 @@
+; ----------------------------------------------------------------------------
+; Literals and comments
+
+ (integer) @number
+ (exp_negation) @number
+ (exp_literal (number)) @float
+ (char) @character
+ [
+   (string)
+   (triple_quote_string)
+ ] @string
+
+ (comment) @comment
+
+
+; ----------------------------------------------------------------------------
+; Punctuation
+
+ [
+   "("
+   ")"
+   "{"
+   "}"
+   "["
+   "]"
+ ] @punctuation.bracket
+
+ [
+ (comma)
+ ";"
+ ] @punctuation.delimiter
+
+
+; ----------------------------------------------------------------------------
+; Keywords, operators, includes
+
+ ; This needs to come before the other "else" in
+ ; order to be highlighted correctly
+ (class_instance "else" @keyword)
+
+ [
+   "if"
+   "then"
+   "else"
+   "case"
+   "of"
+ ] @conditional
+
+ [
+   "import"
+   "module"
+ ] @include
+
+ [
+   (operator)
+   (constructor_operator)
+   (type_operator)
+   (qualified_module)  ; grabs the `.` (dot), ex: import System.IO
+   (all_names)
+
+   ; `_` wildcards in if-then-else and case-of expressions,
+   ; as well as record updates and operator sections
+   (wildcard)
+   "="
+   "|"
+   "::"
+   "∷"
+   "=>"
+   "⇒"
+   "<="
+   "⇐"
+   "->"
+   "→"
+   "<-"
+   "←"
+   "\\"
+   "`"
+   "@"
+ ] @operator
```

Comment:
Some of these might be `@punctuation.delimiter` or `@punctuation.special`.
`(wildcard)` should be `@parameter`.

Reply:
Almost all of them are used specifically as operators, and I don't really see which ones would be a good fit for `@punctuation.*`, except for `@` perhaps. Could you elaborate on which ones you have in mind?

`(wildcard)` has a bit of an unfortunate name, it is actually a lambda constructor, similar to `=>` in JS, so I wouldn't call it a parameter. Would you suggest something else for it?

Reply:
The dot is definitely a delimiter. "::" and "∷" could be as well.
"\\\\" and "`" might be either delimiters or special depending on their usage.

Reply:
Re dot: agreed, will fix.

I think special-casing "∷" relative to operators would look weird in source code. PureScript would often be formatted like this:
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/120750161/6aad7d78-4b6a-43e7-bcd2-29f0dd232703)

`\` is used to introduce a lambda: `\a -> expr`, compare it with JS' `a => expr`. `` ` `` is used to make a function a named infix operator: `add 1 2` would be "regular" notation, and ``1 `add` 2`` is the infix notation with "`".

Wildcards can be used in different contexts:
```purescript
f = case _ of -- lambda-case, the wildcard creates an anonymous function
  x → x

f = if _ then _ else _  -- this is a 3-argument anonymous function
f = if _ then 1 else 2  -- 1-argument anonymous function, etc...

f = (_ + 2)
f = (2 `add` _)         -- operator sections use wildcards to create lambdas

f = _.field             -- anonymous function which accesses field "field" of a record

f = { field: _ }        -- anonymous function which puts a value into field "field"

f = _ { field = _ }     -- two-argument lambda which updates the field "field"
```

Reply:
"∷" can remain an operator.
"\\\\" and "`" could be `@punctuation.special`.
`(wildcard)` is definitely `@parameter`.

Reply:
After thinking about it more, I'm strictly against doing anything to `` ` ``. Consider the examples above (highlighted by Github): this is the classic representation that's been in use for many years, both by PureScript and Haskell and which is based on ergonomics. The function in infix form serves the role of an operator and as such is highlighted as an operator, where the ticks, highlighted the same color, visually extend it and make it easily recognizable. I ran a number of Vim themes with ticks being `@punctuation.special`, but with this change it feels like these ergonomics are gone and the ticks only serve a distracting role.

Wrt the backslash, it makes no sense to edit it without also editing the `->` arrow, since they work together to create a visually distinguishable structure (`\arg -> expr`). But then to preserve the coherence we'd also need to edit `->` in case expression and probably special-case it as a type operator and in functional dependencies. If it's special-cased as a type operator it'd look really weird, on top of not making sense semantically. I am absolutely sure 99% of PureScript users will not be happy about it at all.

Wildcards being `@parameter` seems to be a good suggestion, I will add it.

Reply:
Would also agree than infix ` makes sense as operator. This is also how it is currently handled for Haskell https://github.com/nvim-treesitter/nvim-treesitter/blob/96f55f304332ca6ea1b7dde32d3ec04b5298c316/queries/haskell/highlights.scm#L122